autorest/validations.go

```diff
@@ -0,0 +1,359 @@
+package autorest
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+)
+
+// Constraint stroes constraint name, target field name
+// Rule and chain validations.
+type Constraint struct {
+
+	// Target Field name for validation.
+	Target string
+
+	// Constraint name .e.g. Minlength, Maxlength, Pattern, etc.
+	Name string
+
+	// Rule for constraint .e.g. greater than 10, less than 5 etc.
+	Rule interface{}
+
+	// Chain Validations for struct type
+	Chain *[]Constraint
+}
+
+// Validation stores parameter-wise validation.
+type Validation struct {
+	TargetValue interface{}
+	Constraints []Constraint
+}
+
+const (
+	empty            = "Empty"
+	null             = "Null"
+	readOnly         = "ReadOnly"
+	pattern          = "Pattern"
+	maxLength        = "MaxLength"
+	minLength        = "MinLength"
+	maxItems         = "MaxItems"
+	minItems         = "MinItems"
+	multipleOf       = "MultipleOf"
+	uniqueItems      = "UniqueItems"
+	inclusiveMaximum = "InclusiveMaximum"
+	inclusiveMinimum = "InclusiveMinimum"
+	exclusiveMaximum = "ExclusiveMaximum"
+	exclusiveMinimum = "ExclusiveMinimum"
+)
+
+// Validate method validates constraints on parameter
+// passed in validation array.
+func Validate(m []Validation) error {
+	for _, item := range m {
+		v := reflect.ValueOf(item.TargetValue)
+		for _, constraint := range item.Constraints {
+			var err error
+			switch v.Kind() {
+			case reflect.Ptr:
+				err = validatePtr(item.TargetValue, constraint)
+			case reflect.String:
+				err = validateString(item.TargetValue, constraint)
+			case reflect.Struct:
+				err = validateStruct(item.TargetValue, constraint)
+			case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+				err = validateInt(item.TargetValue, constraint)
+			case reflect.Float32, reflect.Float64:
+				err = validateFloat(item.TargetValue, constraint)
+			case reflect.Array, reflect.Slice, reflect.Map:
+				err = validateArrayMap(item.TargetValue, constraint)
+			default:
+				err = createError(item.TargetValue, constraint, fmt.Sprintf("unknown type %v", v.Kind()))
+			}
+
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func validateStruct(x interface{}, v Constraint) error {
+	z := reflect.ValueOf(x)
+	f := z.FieldByName(v.Target)
```

Comment:
how do you handle if `f` is zero (i.e. when the field with this name is not found)?